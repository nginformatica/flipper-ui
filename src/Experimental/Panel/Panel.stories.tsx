import { Meta, StoryFn } from '@storybook/react'
import React from 'react'
import Panel from '.'

export default {
    title: 'Experimental/Panel',
    component: Panel,
    args: {
        expanded: true,
        title: 'Foo',
        details: <>Panel details</>
    }
} as Meta<typeof Panel>

const Template: StoryFn<typeof Panel> = args => <Panel {...args} />

export const Default = Template.bind({})

export const Nested = Template.bind({})
Nested.args = {
    nested: true
}

export const WithOmittedIcon = Template.bind({})
WithOmittedIcon.args = {
    hideExpansionIcon: true
}
