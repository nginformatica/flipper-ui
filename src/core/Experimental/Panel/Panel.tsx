import { defaultTo } from 'ramda'
import React, { CSSProperties, ChangeEvent, ReactNode } from 'react'
import styled from 'styled-components'
import { ExpandMore as IconExpand } from '../../../icons'
import ExpansionPanel from '../../ExpansionPanel'
import Typography from '../../Typography'

const Paper = styled(ExpansionPanel)`
    && {
        border-radius: 4px;
        ::before {
            height: 0px;
            background-color: transparent;
        }
    }
`

export interface IProps {
    expanded?: boolean
    defaultExpanded?: boolean
    hideSummary?: boolean
    elevation?: number
    summary?: ReactNode
    title?: string | ReactNode
    area?: string
    details?: JSX.Element
    actions?: JSX.Element
    nested?: boolean
    id?: string
    style?: CSSProperties
    className?: string
    hideExpansionIcon?: boolean
    detailsStyle?: CSSProperties
    onChange?(
        event?: ChangeEvent<Record<string, unknown>>,
        expanded?: boolean
    ): void
    onClick?(event: React.MouseEvent<HTMLDivElement>): void
}

const Summary = styled.div`
    display: flex;
    flex: 1;
    align-items: center;
`

const Title = styled(Typography)`
    && {
        flex: 1;
    }
`

const NESTED_ELEVATION = 0
const DEFAULT_ELEVATION = 2

const Panel = (props: IProps) => {
    const {
        hideExpansionIcon,
        expanded,
        defaultExpanded,
        id,
        nested,
        elevation,
        hideSummary,
        title,
        summary,
        details,
        actions,
        area,
        style,
        className,
        detailsStyle,
        onChange,
        onClick,
        ...otherProps
    } = props
    const omitCursor = hideExpansionIcon
        ? { summaryStyle: { cursor: 'auto' } }
        : undefined

    return (
        <Paper
            {...otherProps}
            expanded={expanded}
            id={id}
            data-testid='panel-component'
            defaultExpanded={defaultTo(true, defaultExpanded)}
            elevation={
                nested
                    ? NESTED_ELEVATION
                    : defaultTo(DEFAULT_ELEVATION, elevation)
            }
            summary={
                !hideSummary && (
                    <Summary>
                        {title && (
                            <Title noWrap variant='subtitle1'>
                                {title}
                            </Title>
                        )}
                        {summary}
                    </Summary>
                )
            }
            details={details}
            expandIcon={
                !hideExpansionIcon && <IconExpand data-testid='expand-icon' />
            }
            actions={actions}
            margin='0px'
            style={{
                gridArea: area,
                border: nested ? '1px solid lightgrey' : undefined,
                ...style
            }}
            className={className}
            detailsStyle={detailsStyle}
            onChange={onChange}
            onClick={onClick}
            {...omitCursor}
        />
    )
}

export default Panel
