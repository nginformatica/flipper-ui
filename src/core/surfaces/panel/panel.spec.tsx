import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import Panel from '.'

describe('Panel', () => {
    it('should render', () => {
        render(<Panel title='Foo' details={<button>bar</button>} />)

        const panel = screen.getByText('Foo')
        const expandIcon = screen.getByTestId('expand-icon')

        expect(panel).toBeDefined()
        expect(expandIcon).toBeDefined()
    })

    it('should render with default style', () => {
        render(<Panel title='Foo' details={<button>bar</button>} />)

        const panelContainer = screen.getByTestId('panel-component')

        expect(panelContainer).toHaveProperty('style.border', '')
    })

    it('should render nested', () => {
        render(<Panel nested title='Foo' details={<button>bar</button>} />)

        const panel = screen.getByTestId('panel-component')

        expect(panel).toHaveProperty('style.border', '1px solid lightgrey')
    })

    it('should render with omitted icon', () => {
        render(
            <Panel
                hideExpansionIcon
                title='Foo'
                details={<button>bar</button>}
            />
        )

        waitFor(() => {
            const panel = screen.getByTestId('expand-icon')

            expect(panel).not.toBeDefined()
        })
    })

    it('should match snapshot', () => {
        const { container } = render(
            <Panel title='Foo' details={<button>bar</button>} />
        )

        expect(container).toMatchSnapshot()
    })
})
