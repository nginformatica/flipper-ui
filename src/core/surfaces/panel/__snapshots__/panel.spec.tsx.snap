// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiAccordion-root MuiAccordion-rounded Mui-expanded MuiAccordion-gutters sc-gsFSjX JQcAs css-1lienkd-MuiPaper-root-MuiAccordion-root"
    data-testid="panel-component"
    style="margin: 0px;"
  >
    <div
      aria-expanded="true"
      class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded MuiAccordionSummary-gutters css-1f773le-MuiButtonBase-root-MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content Mui-expanded MuiAccordionSummary-contentGutters css-eqpfi5-MuiAccordionSummary-content"
      >
        <div
          class="sc-iGgVNO iPPrNl"
        >
          <div
            class="sc-kAycRU FJrwN"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-noWrap sc-imWZod cQNAri css-of350e-MuiTypography-root"
            >
              Foo
            </h6>
          </div>
        </div>
      </div>
      <div
        class="MuiAccordionSummary-expandIconWrapper Mui-expanded css-yw020d-MuiAccordionSummary-expandIconWrapper"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="expand-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-pwcg7p-MuiCollapse-root"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
        >
          <div
            class="MuiAccordion-region"
            role="region"
          >
            <div
              class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
            >
              <button>
                bar
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
